src/Lean/Meta/IndPredBelow.lean

```diff
@@ -374,6 +378,184 @@ where
+    - it is an `as` pattern. Here the contstructor could be hidden inside of it.
+    - it is a variable. Here you we need to introduce a fresh variable of a different type.
+  -/
+  convertToBelow (indName : Name) 
+      (originalPattern : Pattern) : MetaM $ Array LocalDecl × Pattern := do 
+    match originalPattern with
+    | Pattern.ctor ctorName us params fields => 
+      let ctorInfo ← getConstInfoCtor ctorName
+      
+      let belowCtor ← getConstInfoCtor $ ctorName.updatePrefix $ ctorInfo.induct ++ `below
+      let belowIndices ← IndPredBelow.getBelowIndices ctorName
+      let belowIndices := belowIndices[ctorInfo.numParams:].toArray.map (· - belowCtor.numParams)
+
+      let mut belowFieldOpts := mkArray belowCtor.numFields none
+      let fields := fields.toArray
+      for fieldIdx in [:fields.size] do
+        belowFieldOpts := belowFieldOpts.set! belowIndices[fieldIdx] (some fields[fieldIdx])
+
+      let belowParams := params.toArray.push belowMotive
+      let belowCtorExpr := mkAppN (mkConst belowCtor.name us) belowParams
+      let (additionalFVars, belowFields) ← transformFields belowCtorExpr indName belowFieldOpts 
+      
+      withExistingLocalDecls additionalFVars.toList do
+      let ctor := Pattern.ctor belowCtor.name us belowParams.toList belowFields.toList
+      trace[Meta.IndPredBelow.match] "{originalPattern.toMessageData} ↦ {ctor.toMessageData}"
+      (additionalFVars, ctor) 
+    | Pattern.as varId p => 
+      let (additionalFVars, p) ← convertToBelow indName p
+      (additionalFVars, Pattern.as varId p)
+    | Pattern.var varId => 
+      let var := mkFVar varId
+      (←inferType var).withApp fun ind args => do
+      let (_, tgtType) ← belowType belowMotive #[var] 0
+      withLocalDeclD (←mkFreshUserName `h) tgtType fun h => do
+      let localDecl ← getFVarLocalDecl h
+      (#[localDecl], Pattern.var h.fvarId!)
+    | p => (#[], p)  
+
+  transformFields belowCtor indName belowFieldOpts := 
+    let rec loop
+      (belowCtor : Expr) 
+      (belowFieldOpts : Array $ Option Pattern)
+      (belowFields : Array Pattern) 
+      (additionalFVars : Array LocalDecl) : MetaM (Array LocalDecl × Array Pattern) := do
+      if belowFields.size ≥ belowFieldOpts.size then (additionalFVars, belowFields) else
+      if let some belowField := belowFieldOpts[belowFields.size] then 
+        let belowFieldExpr ← belowField.toExpr
+        let belowCtor := mkApp belowCtor belowFieldExpr
+        let patTy ← inferType belowFieldExpr
+        patTy.withApp fun f args => do
+        let constName := f.constName?
+        if constName == indName then
+          let (fvars, transformedField) ← convertToBelow indName belowField
+          withExistingLocalDecls fvars.toList do
+          let belowFieldOpts := belowFieldOpts.set! (belowFields.size + 1) transformedField
+          let belowField := 
+            match belowField with 
+            | Pattern.ctor _ _ _ _ => Pattern.inaccessible belowFieldExpr 
```

Comment:
We can use `| Pattern.ctor .. => ` to avoid the `_ ... _`.
It is not relevant for this particular pattern, but the `..` is also quite useful when combined with named parameters.